File: IsometricRPG/IsometricRPG/Entities/Player.swift
import SpriteKit

enum BuffType: CustomStringConvertible {
    case speed
    case damage
    case shield
    case fireRate
    case xpBoost

    var description: String {
        switch self {
        case .speed:    return "speed"
        case .damage:   return "damage"
        case .shield:   return "shield"
        case .fireRate: return "fireRate"
        case .xpBoost:  return "xpBoost"
        }
    }
}

struct ActiveBuff {
    let type: BuffType
    let multiplier: CGFloat
    var remainingDuration: TimeInterval
}

final class Player: Entity {
    var moveDirection: CGPoint = .zero
    var aimDirection: CGPoint = .zero
    var isShooting: Bool = false
    private var lastFireTime: TimeInterval = 0

    // RPG stats
    var level: Int = 1
    var experience: Int = 0
    var experienceToNext: Int = 100
    var armor: Int = 0
    var keys: Int = 0

    // Buffs
    private(set) var activeBuffs: [ActiveBuff] = []

    init(worldPosition: CGPoint) {
        let node = Player.createPlayerNode()
        super.init(worldPosition: worldPosition, health: Constants.playerMaxHealth, node: node)
        node.zPosition = Constants.ZPosition.entity
        node.name = "player"
    }

    private static func createPlayerNode() -> SKNode {
        let container = SKNode()

        let body = SKShapeNode(rectOf: Constants.playerSize, cornerRadius: 3)
        body.fillColor = SKColor(red: 0.2, green: 0.5, blue: 0.9, alpha: 1)
        body.strokeColor = SKColor(red: 0.1, green: 0.3, blue: 0.7, alpha: 1)
        body.lineWidth = 1.5
        container.addChild(body)

        let indicator = SKShapeNode(circleOfRadius: 3)
        indicator.fillColor = .white
        indicator.strokeColor = .clear
        indicator.position = CGPoint(x: 0, y: Constants.playerSize.height / 2 + 4)
        indicator.name = "aimIndicator"
        container.addChild(indicator)

        let shadow = SKShapeNode(ellipseOf: CGSize(width: 18, height: 8))
        shadow.fillColor = SKColor.black.withAlphaComponent(0.3)
        shadow.strokeColor = .clear
        shadow.position = CGPoint(x: 0, y: -Constants.playerSize.height / 2)
        shadow.zPosition = Constants.ZPosition.shadow - Constants.ZPosition.entity
        container.addChild(shadow)

        return container
    }

    override func update(deltaTime: TimeInterval) {
        if moveDirection != .zero {
            let speedMultiplier = buffMultiplier(for: .speed)
            let speed = Constants.playerSpeed * CGFloat(deltaTime) * speedMultiplier
            worldPosition = CGPoint(
                x: worldPosition.x + moveDirection.x * speed / Constants.tileWidth,
                y: worldPosition.y + moveDirection.y * speed / Constants.tileHeight
            )
        }

        if aimDirection != .zero {
            let angle = atan2(aimDirection.y, aimDirection.x) - .pi / 2
            if let indicator = node.childNode(withName: "aimIndicator") {
                let radius: CGFloat = Constants.playerSize.height / 2 + 4
                indicator.position = CGPoint(
                    x: cos(angle + .pi / 2) * radius,
                    y: sin(angle + .pi / 2) * radius
                )
            }
        }

        node.zPosition = Constants.ZPosition.entity + (worldPosition.x + worldPosition.y) * 0.01
        super.update(deltaTime: deltaTime)
    }

    // MARK: - Combat

    func canFire(currentTime: TimeInterval) -> Bool {
        let rateMultiplier = buffMultiplier(for: .fireRate)
        let effectiveRate = Constants.fireRate * TimeInterval(rateMultiplier)
        return isShooting && aimDirection != .zero && (currentTime - lastFireTime) >= effectiveRate
    }

    func didFire(at time: TimeInterval) {
        lastFireTime = time
    }

    func effectiveDamage() -> Int {
        let mult = buffMultiplier(for: .damage)
        return Int(CGFloat(Constants.bulletDamage) * mult)
    }

    override func takeDamage(_ amount: Int) {
        var reduced = amount
        let shieldMult = buffMultiplier(for: .shield)
        if shieldMult < 1.0 {
            reduced = Int(CGFloat(amount) * shieldMult)
        }
        if armor > 0 {
            let absorbed = min(armor, reduced / 2)
            armor -= absorbed
            reduced -= absorbed
        }
        super.takeDamage(max(reduced, 1))
    }

    // MARK: - Healing & Items

    func heal(amount: Int) {
        health = min(health + amount, maxHealth)

        let label = SKLabelNode(text: "+\(amount)")
        label.fontName = "Helvetica-Bold"
        label.fontSize = 12
        label.fontColor = .green
        label.position = CGPoint(x: 0, y: 25)
        label.zPosition = 50
        node.addChild(label)
        let rise = SKAction.moveBy(x: 0, y: 15, duration: 0.6)
        let fade = SKAction.fadeOut(withDuration: 0.4)
        label.run(SKAction.sequence([SKAction.group([rise, fade]), SKAction.removeFromParent()]))
    }

    func addArmor(_ amount: Int) {
        armor += amount
    }

    func addKey() {
        keys += 1
    }

    // MARK: - Buff System

    func applyBuff(_ type: BuffType, duration: TimeInterval, multiplier: CGFloat) {
        activeBuffs.removeAll { $0.type == type }
        node.childNode(withName: "buff_\(type)")?.removeFromParent()

        activeBuffs.append(ActiveBuff(type: type, multiplier: multiplier, remainingDuration: duration))

        let indicator = SKShapeNode(circleOfRadius: 12)
        indicator.fillColor = buffColor(type).withAlphaComponent(0.2)
        indicator.strokeColor = buffColor(type)
        indicator.lineWidth = 1
        indicator.name = "buff_\(type)"
        indicator.zPosition = -0.2
        node.addChild(indicator)

        let pulse = SKAction.sequence([
            SKAction.scale(to: 1.3, duration: 0.4),
            SKAction.scale(to: 0.9, duration: 0.4)
        ])
        indicator.run(SKAction.repeatForever(pulse))
    }

    func updateBuffs(deltaTime: TimeInterval) {
        var expired: [BuffType] = []
        for i in 0..<activeBuffs.count {
            activeBuffs[i].remainingDuration -= deltaTime
            if activeBuffs[i].remainingDuration <= 0 {
                expired.append(activeBuffs[i].type)
            }
        }
        for type in expired {
            node.childNode(withName: "buff_\(type)")?.removeFromParent()
        }
        activeBuffs.removeAll { $0.remainingDuration <= 0 }
    }

    func buffMultiplier(for type: BuffType) -> CGFloat {
        if let buff = activeBuffs.first(where: { $0.type == type }) {
            return buff.multiplier
        }
        return 1.0
    }

    var activeBuffCount: Int { activeBuffs.count }

    private func buffColor(_ type: BuffType) -> SKColor {
        switch type {
        case .speed:    return .cyan
        case .damage:   return .orange
        case .shield:   return .blue
        case .fireRate: return .yellow
        case .xpBoost:  return .purple
        }
    }

    // MARK: - Experience

    func gainExperience(_ amount: Int) {
        let mult = buffMultiplier(for: .xpBoost)
        let total = Int(CGFloat(amount) * mult)
        experience += total
        while experience >= experienceToNext {
            levelUp()
        }
    }

    private func levelUp() {
        level += 1
        experience -= experienceToNext
        experienceToNext = Int(Double(experienceToNext) * 1.5)
        maxHealth += 10
        health = maxHealth

        let label = SKLabelNode(text: "LEVEL UP!")
        label.fontName = "Helvetica-Bold"
        label.fontSize = 14
        label.fontColor = .yellow
        label.position = CGPoint(x: 0, y: 30)
        label.zPosition = 50
        node.addChild(label)

        let rise = SKAction.moveBy(x: 0, y: 20, duration: 1.0)
        let fade = SKAction.fadeOut(withDuration: 1.0)
        label.run(SKAction.sequence([SKAction.group([rise, fade]), SKAction.removeFromParent()]))
    }
}

File: IsometricRPGMono/Entities/Player.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace IsometricRPG;

public enum BuffType { Speed, Damage, Shield, FireRate, XpBoost }

public struct ActiveBuff
{
    public BuffType Type       { get; set; }
    public float    Multiplier { get; set; }
    public double   Duration   { get; set; }
    public double   Remaining  { get; set; }
}

public class Player : Entity
{
    // Inventory: 20 nullable slots
    public List<Item?> Inventory  { get; } = new List<Item?>(new Item?[20]);
    public EquipmentSet Equipment  { get; } = new EquipmentSet();

    public int    Level          { get; private set; } = 1;
    public double XP             { get; private set; } = 0;
    public double XPToNextLevel  => 100.0 * Math.Pow(1.5, Level - 1);

    private double _lastFireTime = 0;
    private readonly List<ActiveBuff> _buffs = new List<ActiveBuff>();

    public IReadOnlyList<ActiveBuff> ActiveBuffs => _buffs;
    public int ActiveBuffCount => _buffs.Count;

    public Player(Vector2 worldPosition)
        : base(worldPosition, Constants.PlayerMaxHealth) { }

    // ---- Buff System ----

    public void ApplyBuff(BuffType type, float multiplier, double duration)
    {
        _buffs.RemoveAll(b => b.Type == type);
        _buffs.Add(new ActiveBuff { Type = type, Multiplier = multiplier,
                                    Duration = duration, Remaining = duration });
    }

    public void UpdateBuffs(double deltaTime)
    {
        for (int i = 0; i < _buffs.Count; i++)
        {
            var b = _buffs[i];
            b.Remaining -= deltaTime;
            _buffs[i] = b;
        }
        _buffs.RemoveAll(b => b.Remaining <= 0);
    }

    /// Returns the product of all active multipliers for the given buff type (1.0 if none).
    public float BuffMultiplier(BuffType type)
    {
        float product = 1f;
        foreach (var b in _buffs)
            if (b.Type == type) product *= b.Multiplier;
        return product;
    }

    // ---- Combat ----

    public bool CanFire(double currentTime)
    {
        double rateMultiplier = BuffMultiplier(BuffType.FireRate);
        double effectiveRate  = Constants.FireRate * rateMultiplier;
        return (currentTime - _lastFireTime) >= effectiveRate;
    }

    public void DidFire(double currentTime)
    {
        _lastFireTime = currentTime;
    }

    public float EffectiveDamage() => Constants.BulletDamage * BuffMultiplier(BuffType.Damage);

    public float EffectiveSpeed() => Constants.PlayerSpeed * BuffMultiplier(BuffType.Speed);

    // ---- XP / Level ----

    public void AddXP(double amount)
    {
        double boosted = amount * BuffMultiplier(BuffType.XpBoost);
        XP += boosted;
        while (XP >= XPToNextLevel)
        {
            XP -= XPToNextLevel;
            Level++;
            MaxHealth += 10;
            Health = MaxHealth;
        }
    }

    // ---- Inventory helpers ----

    public bool AddToInventory(Item item)
    {
        for (int i = 0; i < Inventory.Count; i++)
        {
            if (Inventory[i] == null)
            {
                Inventory[i] = item;
                return true;
            }
        }
        return false;
    }

    // ---- Update / Draw ----

    public override void Update(GameTime gameTime)
    {
        base.Update(gameTime);
        UpdateBuffs(gameTime.ElapsedGameTime.TotalSeconds);
    }

    /// Move player in world-space using the pre-computed direction and deltaTime.
    /// Wall-slide collision is handled in GameScene.
    public void Move(Vector2 direction, double deltaTime)
    {
        if (direction.LengthSquared() < 1e-6f) return;
        float speed  = EffectiveSpeed();
        WorldPosition += direction * speed * (float)deltaTime / Constants.TileWidth;
    }

    public override void Draw(SpriteBatch spriteBatch, SpriteManager spriteManager, Vector2 cameraPos)
    {
        var screenPos = ScreenPosition - cameraPos;
        var size = Constants.PlayerSize;

        // Body rectangle
        var tex = spriteManager.GetRect(
            (int)size.X,
            (int)size.Y,
            TintColor == Color.White ? new Color(70, 130, 180) : TintColor);  // steel blue

        spriteBatch.Draw(
            tex,
            new Vector2(screenPos.X - size.X / 2f, screenPos.Y - size.Y / 2f),
            Color.White);

        // Simple health bar
        DrawHealthBar(spriteBatch, spriteManager, screenPos, size);
    }

    private void DrawHealthBar(SpriteBatch sb, SpriteManager sm, Vector2 screenPos, Vector2 size)
    {
        const int barW = 24, barH = 4;
        float ratio = Math.Max(0f, (float)Health / MaxHealth);
        var bgTex   = sm.GetRect(barW, barH, new Color(60, 60, 60));
        var fgColor = ratio > 0.5f ? Color.Green : (ratio > 0.25f ? Color.Yellow : Color.Red);
        var fgTex   = sm.GetRect(Math.Max(1, (int)(barW * ratio)), barH, fgColor);

        var barOrigin = new Vector2(screenPos.X - barW / 2f, screenPos.Y - size.Y / 2f - 8f);
        sb.Draw(bgTex, barOrigin, Color.White);
        if (ratio > 0f)
            sb.Draw(fgTex, barOrigin, Color.White);
    }
}
